### DISCUSSION TITLE
Keyboard shortcuts for assigning flags

### DISCUSSION THREAD

**Discussion by ghylander** [wkentaro/labelme#1062] [Q&A / Help]
**Is your feature request related to a problem? Please describe.**
I'd like to be able to create shortcuts to quickly start a polygon tool or assign a flag to the image,  and to advance to the next image or to go back to the previous image. 

**Describe the solution you'd like**
Either pre-defined or user-definable keyboard shortcuts for labelme functions

**Comment by wkentaro** [comment]
You can edit `~/.labelmerc` file and add shortcuts there.
Here are predefined ones:

https://github.com/wkentaro/labelme/blob/f8097a4b19c79ca3328022cfbc351f2c7bff20d5/labelme/config/default_config.yaml#L71-L106

**Comment by ghylander** [comment]
that's amazing to know, I had a look at the docs/tried hitting some keys but did not try 'ctrl+key' combinations
Should i close it or do you prefer to keep it open until docs are ready?

**Comment by ghylander** [comment]
I've been reading the source code to try and add a shortcut to quickly assign flags, but I'm having some trouble

If I understood it correctly, I have to add the shortcut names to and key combinations to the `~/.labelmerc` file:
```
  assign_flag_1: P
  assign_flag_2: N
```

Then, in app.py, I have to create the action(s), which I have initially written as:
```
        assign_flag_1 = action(
            self.tr("&Assign flag"),
            self.assignFlag(1),
            shortcuts["assign_flag_1"],
            "edit",
            self.tr("Assigns flag 1 to the current image"),
        )
        assign_flag_2 = action(
            self.tr("&Assign flag"),
            self.assignFlag(2),
            shortcuts["assign_flag_2"],
            "edit",
            self.tr("Assigns flag 2 to the current image"),
        )
```

And in app.py as well, I define the assignFlag function, `def assignFlag(self, flag_n):`, and this is where I got stuck. I've been trying to read the existing functions (such as editLabel), but I'm having trouble understanding it.

**Reply by SizheAn** [reply]
> I've been reading the source code to try and add a shortcut to quickly assign flags, but I'm having some trouble
> 
> If I understood it correctly, I have to add the shortcut names to and key combinations to the `~/.labelmerc` file:
> 
> ```
>   assign_flag_1: P
>   assign_flag_2: N
> ```
> 
> Then, in app.py, I have to create the action(s), which I have initially written as:
> 
> ```
>         assign_flag_1 = action(
>             self.tr("&Assign flag"),
>             self.assignFlag(1),
>             shortcuts["assign_flag_1"],
>             "edit",
>             self.tr("Assigns flag 1 to the current image"),
>         )
>         assign_flag_2 = action(
>             self.tr("&Assign flag"),
>             self.assignFlag(2),
>             shortcuts["assign_flag_2"],
>             "edit",
>             self.tr("Assigns flag 2 to the current image"),
>         )
> ```
> 
> And in app.py as well, I define the assignFlag function, `def assignFlag(self, flag_n):`, and this is where I got stuck. I've been trying to read the existing functions (such as editLabel), but I'm having trouble understanding it.

Hi, any luck in shortcuts for assigning the flags?

**Comment by owardlaw** [comment]
Having the same issue

**Comment by jeremia49** [comment]
I've tried to implement the shortcut for assigning flag, you can look at my commits below :
https://github.com/jeremia49/labelme/commit/d9ac5d8b422bddb2a6f79500e4ca654301f60d22

I wonder why I can't use numeric keys for the shortcut key so I changed the config file to use alphabetical keys and it works.
https://github.com/jeremia49/labelme/commit/06380c954875ad54b2b0fc3bd5f130327206cde9

**Reply by agentmorris** [reply]
In case it's helpful to the next person... I was also curious why numeric shortcuts didn't work... alt+0 and ctrl+0 worked fine, for example, but just "0" did not work.  I don't have a great explanation, but I have a fix; in qt.py, I explicitly cast integer-valued shortcut strings to QtKey codes, by changing this:

```
if shortcut is not None:
        if isinstance(shortcut, (list, tuple)):
            a.setShortcuts(shortcut)
        else:
            a.setShortcut(shortcut)
```
...to this:

```
if shortcut is not None:
        try:
            val = int(shortcut)
            assert val >= 0 and val <=9, 'Invalid numeric shortcut: {}'.format(shortcut)
            shortcut = QKeySequence(val+QtCore.Qt.Key_0)
        except Exception:
            pass
        if isinstance(shortcut, (list, tuple)):
            a.setShortcuts(shortcut)
        else:
            a.setShortcut(shortcut)
```

...and voila, fixed.
